fix(aio): show embedded ToC (and other fixes)

gkalpak opened the pull request:
The main thing this PR fixes is showing the embedded ToC in narrow screens.

On narrow screens (where there is not enough room on the right to show the floating ToC), an embedded ToC is shown (via an `<aio-toc embedded>` element in the document). Since ToC was not a custom element, the component was not instantiated for the embedded element.

This PR fixes it by making `aio-toc` a custom element and loading it manually for the floating ToC (if necessary).

gkalpak committed:
fix(aio): do not load custom elements again while already loading

gkalpak committed:
refactor(aio): order custom elements by selector

gkalpak committed:
feat(aio): add component for lazy-loading custom element

gkalpak committed:
fix(aio): show embedded ToC

On narrow screens (where there is not enough room on the right to show
the floating ToC), an embedded ToC is shown (via an `<aio-toc embedded>`
element in the document). Since ToC was not a custom element, the
component was not instantiated for the embedded element.

This commit fixes it by making `aio-toc` a custom element and loading it
manually for the floating ToC (if necessary).

gkalpak committed:
fix(aio): avoid loading the ToC component until it is necessary

mary-poppins commented:
You can preview e6204c6 at https://pr23944-e6204c6.ngbuilds.io/.

petebacondarwin commented on aio/src/app/custom-elements/elements-loader.ts:
```diff
@@ -25,30 +27,56 @@ export class ElementsLoader {
-    return this.moduleFactoryLoader.load(modulePath).then(elementModuleFactory => {
-      if (!this.elementsToLoad.has(selector)) { return; }
+  /** Loads and registers the custom element defined on the `WithCustomElement` module factory. */
+  loadCustomElement(selector: string): Promise<void> {
```
Why is this no longer private? Is it just to aid testing?

gkalpak replied:
It is for calling it from [LazyCustomElementComponent][1]. But (given I don't like `LazyCustomElementComponent` anyway) maybe it is better to have a `CustomElementContainer` component that just calls `loadContainingCustomElements()` on itself.

[1]: https://github.com/angular/angular/pull/23944/commits/1a607e011ac9d9795e82cde5b923ab0cc3e78cca#diff-92f5584a0ad28a4ebbebc9bae41d26f0

gkalpak replied:
To expand a bit:
Instead of having a component (`LazyCustomElement`) that calls `loadCustomElement()` to load a specific custom element, there would be a component(/directive?) (`CustomElementContainer`) that would call `loadContainingCustomElements()` on itself.

The reason I went with the `LoadCustomElement` component initially was performance (i.e. not having to go through the element content and extract custom elements) each time, but this is probably not a huge concern (since the contents of `CustomElementContainer` will generally be short and shallow and we only look for elements not yet loaded).

Also, with the `LoadCustomElement` component, extra effort and complexity are required in order to support passing attributes and content to the custom elements (currently not implemented). With a `CustomElementContainer`-like component, it should be supported out of the box.

petebacondarwin replied:
I agree that the container component approach is better.
I guess the downside is that if there are some custom elements that are never loaded, due to never being used on a page, then this container component will always have to query its content for those elements.
Perhaps it would be possible to cache the results of `loadContainingCustomElements` for each instance of this container component somehow?

petebacondarwin commented on aio/src/app/custom-elements/elements-loader.ts:
```diff
@@ -25,30 +27,56 @@ export class ElementsLoader {
+    }
+
+    // The custom element has already been loaded and registered.
+    return Promise.resolve();
```
Should this code path ever be hit? It seems that if a custom element has been loaded then it has been removed from the `elementsToLoad` collection so this method would not have been called.

Unless, now that it is public something else might call this?

gkalpak replied:
See https://github.com/angular/angular/pull/23944/files#r189807893.

petebacondarwin replied:
I see - so the container approach would allow this code path to be removed.

petebacondarwin commented on aio/src/app/custom-elements/elements-loader.spec.ts:
```diff
@@ -63,63 +51,213 @@ describe('ElementsLoader', () => {
+  });
 });
 
 // TEST CLASSES/HELPERS
```
Not related to this PR but there are now "helper" classes at both the top and bottom of this file.

gkalpak replied:
Fixed.

petebacondarwin commented on aio/src/app/custom-elements/elements-loader.ts:
```diff
@@ -25,30 +27,56 @@ export class ElementsLoader {
    * elements so that they will not be queried in subsequent calls.
    */
   loadContainingCustomElements(element: HTMLElement): Observable<void> {
```
Not related to this PR but the name of this method makes no sense to me... Shouldn't it be `loadContainedCustomElements`, i.e. the custom elements that are "contained" by this element?

petebacondarwin commented on aio/src/app/custom-elements/elements-loader.spec.ts:
```diff
@@ -63,63 +51,213 @@ describe('ElementsLoader', () => {
 
-    elementsLoader.loadContainingCustomElements(hostEl);
-    tick(); // Tick for the module factory loader's async `load` function
+    // `loadCustomElement()` uses the `window.customElements` API. Provide a mock for this test.
```
This seems rather complicated - why not just spy on the two methods that you care about?

```typescript
spyOn(window.customElements, 'define');
spyOn(window.customElements, 'whenDefined');
```

Then Jasmine will take care of reverting the spy after each test, saving the need for the `afterEach` block.

Or is the `window.customElements` locked down to prevent this?

gkalpak replied:
No, it's not locked. Using spies sounds reasoable; not sure what I was thinking 😕
<sub>(Maybe I was carried away by the [previous implementation][1], manually overwriting the method 😇)</sub>

[1]: https://github.com/angular/angular/pull/23944/files#diff-fc503f29d75fcbafa5044f3928565129L45

gkalpak replied:
Fixed.

petebacondarwin commented on aio/src/app/custom-elements/elements-loader.spec.ts:
```diff
@@ -63,63 +51,213 @@ describe('ElementsLoader', () => {
+
+      // Verify the right component was loaded/registered.
+      const Ctor = mockCustomElements.define.calls.argsFor(0)[1];
+      expect(Ctor.observedAttributes).toEqual(['element-a-input']);
```
It is a shame that `createCustomElement` is a freestanding exported function, otherwise we could spy on it and return a specific object that we could look for here.

petebacondarwin commented on aio/src/app/custom-elements/elements-loader.spec.ts:
```diff
@@ -63,63 +51,213 @@ describe('ElementsLoader', () => {
+      elementsLoader.loadCustomElement('element-a-selector').then(() => state = 'resolved');
+      flushMicrotasks();
+      expect(state).toBe('resolved');
+      expect(mockCustomElements.define).not.toHaveBeenCalled();
```
One could argue that this last situation is actually a programming error, in which case we should error rather than resolve this...

gkalpak replied:
I think it is fine to call this method for an already loaded element.

petebacondarwin replied:
I understand now that this is a valid scenario because I didn't see that it might be called from the lazy custom element component. But if we removed that then this suggestion again stand.

petebacondarwin reviewed (approved):
A few minor comments but generally LGTM.

One thought I had about the behaviour... if any custom elements fail then the whole of the containing element fails, right? if you allow retries of custom elements that failed this would only be for future instances of elements that might contain the originally failing custom element.

How is the user impacted in such a "temporary" failure. Is there some way to tell the user that there was an error and that they might want to try refreshing the page or something?

gkalpak commented on aio/src/app/custom-elements/elements-loader.spec.ts:
```diff
@@ -63,63 +51,213 @@ describe('ElementsLoader', () => {
+
+    beforeEach(() => loadCustomElementSpy = spyOn(elementsLoader, 'loadCustomElement'));
+
+    it('should load and register all contained elements', fakeAsync(() => {
```
Don't be pessimistic 😛

gkalpak replied:
Fixed.

petebacondarwin commented:
Looking again at this PR (now I noticed ALL the commits!?!) I wonder if we need the lazy loading component stuff at all (whether as it is now or via a container component approach)?

The only use case is for lazy loading the floating ToC if it is needed, right?
But since we have direct knowledge of when this is needed (it is set in the `AppComponent`), could we not just trigger this custom element to be loaded programmatically when we first realise that a page would need to display it?

This would have the benefit of not needing to have the additional complexity of this new lazy loading (container) component and the performance benefit of not having to check whether it is loaded inside such a component every time that component is rendered.

gkalpak commented:
Yes, that was the other option I was considering. The reason I didn't go with it was that it is ad-hoc and not re-usable. Using a component to load the custom element hides all knowledge about the `ElementLoader` etc from the consuming component.

But maybe I should worry about re-usability once we actually need the functionality in more than one place :grin:

petebacondarwin commented:
That would be the "agile" approach :-)

gkalpak commented:
Is that an argument for against doing it this way? 😛 
OK, I'll ditch the `LazyCustomElement` component altogether for now 😞

gkalpak committed:
fixup! fix(aio): do not load custom elements again while already loading

gkalpak committed:
refactor(aio): rename method (loadContainingCustomElements --> loadContainedCustomElements)

gkalpak commented:
I pushed two new commits addressing most comments.

##
I also played with the different approaches again and non seems to be ideal.

First of all, it is not possible to use the custom elements in the HTML directly, unless defining dummy components for each (due to "unknown element" errors). And if you don't use them directly in the HTML (but instead inject them via `nativeElement.innerHTML`), you can't use bindings etc. This affects all three approaches (`LazyCustomElement`, `CustomElementContainer`, manually loading the custom element in `AppComponent`).

Each approach has its own pros and cons (and complexities), so `LazyCustomElement` is as good as any for our current requirements. 

##
Maybe we could merge this as is to fix embedded ToC and switch to a different approach once we figure out a better one :smiley:

petebacondarwin commented on aio/src/app/custom-elements/elements-loader.spec.ts:
```diff
@@ -266,11 +231,28 @@ class FakeCustomElementModule implements WithCustomElementComponent {
+         projectableNodes?: any[][],
+         rootSelectorOrNode?: string | any,
+         ngModule?: NgModuleRef<any>): ComponentRef<any> {
+    return (jasmine.createSpy('ComponentRef') as any) as ComponentRef<any>;
```
NIT: Because you have specified a return type for this method you don't need to cast the `any` to the return type.

gkalpak replied:
Fixed.

mary-poppins commented:
You can preview ccd8148 at https://pr23944-ccd8148.ngbuilds.io/.

gkalpak committed:
fixup! fix(aio): do not load custom elements again while already loading

mary-poppins commented:
You can preview a807ef6 at https://pr23944-a807ef6.ngbuilds.io/.

gkalpak committed:
build(aio): update payload size limits

gkalpak committed:
fixup! fix(aio): do not load custom elements again while already loading

gkalpak committed:
fixup! fix(aio): show embedded ToC

gkalpak committed:
fixup! refactor(aio): rename method (loadContainingCustomElements --> loadContainedCustomElements)

mary-poppins commented:
You can preview 5531d20 at https://pr23944-5531d20.ngbuilds.io/.

gkalpak commented:
I added a couple of commits updating the payload size limits, because CI was failing.
(Nothing unexpected there. `runtime.js` went up by ~50B since a new "lazy bundle" is added and `main.js` went ~4KB down since `TocComponent` was moved to a "lazy bundle".)

Once all checks pass, I'll mark it for merge again.

mary-poppins commented:
You can preview 0923da6 at https://pr23944-0923da6.ngbuilds.io/.

mary-poppins commented:
You can preview 38621f3 at https://pr23944-38621f3.ngbuilds.io/.
